### PR TITLE
AUC metrics

### PR DESCRIPTION
**Pull request by liliarose**
## Patch description
Adding AUC (Area under ROC curve) metrics as an option for eval_model

## Testing steps
0. Check the other metric tests r okay (` pytest -v -k TestMetrics/ pytest -v -k TestMetric`)
1. Check via test to make sure AUC added up correctly..... (`pytest -v -k TestAggregators`) 
2. Check it works via `eval_model` for both classes with micro aggregation --> checking to make sure the agents' auc resetted.  (`parlai em -mf zoo:dialogue_safety/multi_turn/model -t internal:civil_bias_toxic_comment:civil_bias_toxicity,dialogue_safety -auc 4 -rf parlai_internal/reports/auc_multi_safety_civil_multi.json -ne 3000 -micro True`)

## Logs
Screen Shot 2021-06-30 at 1 36 01 PM

Screen Shot 2021-06-30 at 1 36 17 PM

metric|dialogue_safety|internal:civil_bias_toxic_comment:civil_bias_toxicity|all
---|---|---|---
class___notok___f1|0.82453|0.55437|0.68945
AUC___notok__|0.98469|0.92527|0.96373

### Testing Step 2
```
parlai em -mf zoo:dialogue_safety/multi_turn/model -t internal:civil_bias_toxic_comment:civil_bias_toxicity -auc True -rf parlai_internal/reports/auc_multi_safety_civil_single.json -ne 3000
```
Results:

### Testing Step 1

```
(conda_parlai) wzhang4343@devfair0173:~/ParlAI$ pytest -v -k TestAggregators
===================================================== test session starts =====================================================
platform linux -- Python 3.7.10, pytest-6.2.4, py-1.10.0, pluggy-1.0.0.dev0 -- /private/home/wzhang4343/.conda/envs/conda_parlai/bin/python
cachedir: .pytest_cache
rootdir: /private/home/wzhang4343/ParlAI, configfile: pytest.ini, testpaths: tests, parlai/tasks
plugins: requests-mock-1.9.2, regressions-2.2.0, hydra-core-1.0.6, datadir-1.3.1
collected 997 items / 990 deselected / 7 selected                                                                             

tests/test_metrics.py::TestAggregators::test_auc_metrics PASSED                                                         [ 14%]
tests/test_metrics.py::TestAggregators::test_classifier_metrics PASSED                                                  [ 28%]
tests/test_metrics.py::TestAggregators::test_macro_aggregation PASSED                                                   [ 42%]
tests/test_metrics.py::TestAggregators::test_micro_aggregation PASSED                                                   [ 57%]
tests/test_metrics.py::TestAggregators::test_time_metric PASSED                                                         [ 71%]
tests/test_metrics.py::TestAggregators::test_uneven_macro_aggrevation PASSED                                            [ 85%]
tests/test_metrics.py::TestAggregators::test_unnamed_aggregation PASSED                                                 [100%]

==================================================== slowest 10 durations =====================================================
0.01s call     tests/test_metrics.py::TestAggregators::test_auc_metrics

(9 durations < 0.005s hidden.  Use -vv to show these durations.)
======================================== 7 passed, 990 deselected, 2 warnings in 5.77s ========================================
```

### Testing Step 0
```
(conda_parlai) wzhang4343@devfair0173:~/ParlAI$ pytest -v -k TestMetric
===================================================== test session starts =====================================================
platform linux -- Python 3.7.10, pytest-6.2.4, py-1.10.0, pluggy-1.0.0.dev0 -- /private/home/wzhang4343/.conda/envs/conda_parlai/bin/python
cachedir: .pytest_cache
rootdir: /private/home/wzhang4343/ParlAI, configfile: pytest.ini, testpaths: tests, parlai/tasks
plugins: requests-mock-1.9.2, regressions-2.2.0, hydra-core-1.0.6, datadir-1.3.1
collected 996 items / 984 deselected / 12 selected                                                                            

tests/test_metrics.py::TestMetric::test_average_metric_additions PASSED                                                 [  8%]
tests/test_metrics.py::TestMetric::test_average_metric_inputs PASSED                                                    [ 16%]
tests/test_metrics.py::TestMetric::test_fixedmetric PASSED                                                              [ 25%]
tests/test_metrics.py::TestMetric::test_macroaverage_additions PASSED                                                   [ 33%]
tests/test_metrics.py::TestMetric::test_sum_metric_additions PASSED                                                     [ 41%]
tests/test_metrics.py::TestMetric::test_sum_metric_inputs PASSED                                                        [ 50%]
tests/test_metrics.py::TestMetrics::test_largebuffer PASSED                                                             [ 58%]
tests/test_metrics.py::TestMetrics::test_multithreaded PASSED                                                           [ 66%]
tests/test_metrics.py::TestMetrics::test_recent PASSED                                                                  [ 75%]
tests/test_metrics.py::TestMetrics::test_shared PASSED                                                                  [ 83%]
tests/test_metrics.py::TestMetrics::test_simpleadd PASSED                                                               [ 91%]
tests/test_metrics.py::TestMetrics::test_verymultithreaded PASSED                                                       [100%]

==================================================== slowest 10 durations =====================================================
0.13s call     tests/test_metrics.py::TestMetrics::test_verymultithreaded
0.09s call     tests/test_metrics.py::TestMetrics::test_largebuffer

(8 durations < 0.005s hidden.  Use -vv to show these durations.)
======================================= 12 passed, 984 deselected, 2 warnings in 7.61s ========================================
(conda_parlai) wzhang4343@devfair0173:~/ParlAI$ pytest -v -k TestMetrics
===================================================== test session starts =====================================================
platform linux -- Python 3.7.10, pytest-6.2.4, py-1.10.0, pluggy-1.0.0.dev0 -- /private/home/wzhang4343/.conda/envs/conda_parlai/bin/python
cachedir: .pytest_cache
rootdir: /private/home/wzhang4343/ParlAI, configfile: pytest.ini, testpaths: tests, parlai/tasks
plugins: requests-mock-1.9.2, regressions-2.2.0, hydra-core-1.0.6, datadir-1.3.1
collected 996 items / 990 deselected / 6 selected                                                                             

tests/test_metrics.py::TestMetrics::test_largebuffer PASSED                                                             [ 16%]
tests/test_metrics.py::TestMetrics::test_multithreaded PASSED                                                           [ 33%]
tests/test_metrics.py::TestMetrics::test_recent PASSED                                                                  [ 50%]
tests/test_metrics.py::TestMetrics::test_shared PASSED                                                                  [ 66%]
tests/test_metrics.py::TestMetrics::test_simpleadd PASSED                                                               [ 83%]
tests/test_metrics.py::TestMetrics::test_verymultithreaded PASSED                                                       [100%]

==================================================== slowest 10 durations =====================================================
0.13s call     tests/test_metrics.py::TestMetrics::test_verymultithreaded
0.09s call     tests/test_metrics.py::TestMetrics::test_largebuffer

(8 durations < 0.005s hidden.  Use -vv to show these durations.)
======================================== 6 passed, 990 deselected, 2 warnings in 5.16s ========================================
```

** Other Info **
- I chose to store the AUCmetric in the classifier because if we ever wanna do it during training.... it should be relatively easy to modify the code and do it there :) (maybe?)
- Also, added 6 tests and know that it might be a lot less code if I just did a for loop with some of the commands, but then I noticed that if I did do that, it would be less easy to read exactly which test it was. 
- Please don't get scared by the lines...... ~300 are in the tests. 
- Code used to generate the graphs (not in the final version) 
```
                # for fun...
                import matplotlib.pyplot as plt
                from sklearn import metrics
                fpr, tpr, _, _ = curr_auc._calc_fpr_tpr()
                display = metrics.RocCurveDisplay(fpr=fpr, tpr=tpr, roc_auc=curr_auc.value())
                display.plot()
                folder = '/'.join(opt['report_filename'].split('/')[:-1])
                plt.savefig(folder+f"/AUC_graph_{task}_{classifier_agent.class_list[class_indices]}.png")
                plt.clf()
                print(f"graphed {folder}/AUC_graph_{task}_{classifier_agent.class_list[class_indices]}.png")
```
- Code used to generate table from report file:

```
interested_metrics = {'AUC___notok__', 'class___notok___f1'}

fields = ['dialogue_safety', 'internal:civil_bias_toxic_comment:civil_bias_toxicity']
rows = []

for metric in interested_metrics:  
 row = [metric]
 for field in fields:
   key = field + '/' + metric 
   row.append(str(round(report[key], 5)))
 row.append(str(round(report[metric], 5)))
 rows.append('|'.join(row))

print('|'.join(['metric'] + fields + ['all']))
print('|'.join(['---']*(len(fields) + 2)))
print('\n'.join(rows))
```

### CODE REVIEW

File: parlai/core/torch_classifier_agent.py [hunk 1]
```diff
@@ -168,6 +170,127 @@ def value(self) -> float:
             return numer / denom
 
 
+class AUCMetrics(Metric):
+    """
+    Class that calculates the area under the roc curve from list of labels and its true
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py:175]: nit: docstrings should have a single sentence intro, then a more detailed description

```
Computes Area Under ROC Curve (AUC) metrics.

Does so by keeping track of positives and negatives. Note the introduction of
`max_bucket_dec_places`, which is used for ...
```

File: parlai/core/torch_classifier_agent.py [hunk 2]
```diff
@@ -168,6 +170,127 @@ def value(self) -> float:
...
+    probabilities; expecting values to be (false positives, true positives)
+    """
+
+    __slots__ = (
+        '_pos_dict',
+        '_tot_pos',
+        '_neg_dict',
+        '_tot_neg',
+        '_class_name',
+        '_max_bucket_dec_places',
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py:185]: I don't quite understand what this is for.

File: parlai/core/torch_classifier_agent.py [hunk 3]
```diff
@@ -168,6 +170,127 @@ def value(self) -> float:
...
+            fp_tp.append((fp, tp))
+        return fp_tp
+
+    def value(self) -> float:
+        _tot_pos = sum(self._pos_dict.values())
+        _tot_neg = sum(self._neg_dict.values())
+        if _tot_pos == 0 and _tot_neg == 0:
+            return 0
+        fp_tp = self._calc_fp_tp()
+        fp_tp.sort(key=lambda x: x[0])
```
Comment (stephenroller) [inline parlai/core/torch_classifier_agent.py:279]: the default for sorting tuples is by the first item in the tuples, with ties broken by the second item in tuples, ... Is that sufficient?

**Review by stephenroller — COMMENTED** [review]
Generally seems good and I appreciate the feature. Just have a clarification question.

**Comment by stephenroller** [comment]
(That test is fantastic btw)

File: parlai/core/torch_classifier_agent.py (hunk 2 above)
Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:185]: According to a post I read.... adding properties/attributes to slots would make it faster to access.... but if it's not necessary, then I'll delete it.

File: parlai/core/torch_classifier_agent.py (hunk 3 above)
Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:279]: ohhhh, I'll check that

File: parlai/core/torch_classifier_agent.py (hunk 2 above)
Comment (stephenroller) [reply parlai/core/torch_classifier_agent.py]: I meant more the specific property of `max_bucket_dec_places`

File: parlai/core/torch_classifier_agent.py [hunk 4]
```diff
@@ -168,6 +170,129 @@ def value(self) -> float:
...
+        """
+        return False
+
+    @classmethod
+    def raw_data_to_auc(
+        cls,
+        true_labels: List[int],
+        pos_probs: List[float],
+        class_name,
+        max_bucket_dec_places: float = 5,
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:197]: nit why is the default `max_bucket_dec_places` different from L210

Comment (jxmsML) [reply parlai/core/torch_classifier_agent.py:197]: also nit: change the type of `max_bucket_dec_places` to `int` might be better?

File: parlai/core/torch_classifier_agent.py [hunk 5]
```diff
@@ -168,6 +170,129 @@ def value(self) -> float:
...
+            true_labels=true_labels, pos_probs=pos_probs, class_name=class_name
+        )
+        return auc_object
+
+    def __init__(
+        self,
+        class_name: Union[int, str],
+        pos_dict: Dict[float, int] = None,
+        neg_dict: Dict[float, int] = None,
+        max_bucket_dec_places: float = 3,
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:210]: same here `max_bucket_dec_places: int`

File: parlai/core/torch_classifier_agent.py (hunk 4 above)
Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:197]: ahhh.... forgot to change; thanks!

File: parlai/core/torch_classifier_agent.py [hunk 6]
```diff
@@ -168,6 +170,129 @@ def value(self) -> float:
...
+        auc_object = cls(class_name, max_bucket_dec_places=max_bucket_dec_places)
+        auc_object.update_raw(
+            true_labels=true_labels, pos_probs=pos_probs, class_name=class_name
+        )
+        return auc_object
+
+    def __init__(
+        self,
+        class_name: Union[int, str],
+        pos_dict: Dict[float, int] = None,
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:208]: Nit: maybe Dict to Count

File: parlai/core/torch_classifier_agent.py [hunk 7]
```diff
@@ -168,6 +170,129 @@ def value(self) -> float:
...
+            # true positives
+            tp += self._pos_dict.get(thres, 0)
+            fp_tp.append((fp, tp))
+        return fp_tp
+
+    def _calc_fpr_tpr(self) -> Tuple[List[int]]:
+        _tot_pos = sum(self._pos_dict.values())
+        _tot_neg = sum(self._neg_dict.values())
+        fp_tp = self._calc_fp_tp()
+        fp_tp.sort()
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:273]: what is this sort list of Tuple doing?

Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:273]: Auc wants the x-axis to be sorted, but I guess makes more sense to sort it inside value instead of here....

File: parlai/scripts/eval_model.py [hunk 8]
```diff
@@ -42,6 +42,8 @@
     get_rank,
 )
 
+CLASSIFIER_AGENT = 1
```
Comment (EricMichaelSmith) [inline parlai/scripts/eval_model.py:45]: A comment here would be quite useful

**Review by EricMichaelSmith — COMMENTED** [review]
Wow, great to have! Yeah, I like the thorough tests. Will defer to others with more context for approval

File: parlai/core/torch_classifier_agent.py [hunk 9]
```diff
@@ -168,6 +171,133 @@ def value(self) -> float:
...
+    ):
+        self._pos_dict = pos_dict if pos_dict else Counter()
+        self._neg_dict = neg_dict if neg_dict else Counter()
+        self._class_name = class_name
+        self._max_bucket_dec_places = max_bucket_dec_places
+
+    def update_raw(
+        self, true_labels: List[Union[int, str]], pos_probs: List[float], class_name
+    ):
+        assert self._class_name == class_name
```
[the remark this answers is not in the document]
Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:221]: Currently, we only have 2 classes, but if there were more classes and we use this for a one-vs rest approach, it would be a security measure to make sure that the classes won't get mixed up for some reason; more comments sounds like a good idea :)

File: parlai/core/torch_classifier_agent.py [hunk 10]
```diff
@@ -15,14 +15,17 @@
...
 from parlai.utils.typing import TScalar
 from parlai.utils.io import PathManager
-import parlai.utils.logging as logging
-
+from sklearn.metrics import auc
 
-import torch
+import parlai.utils.logging as logging
 import torch.nn.functional as F
+from collections import Counter
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:26]: nit is L26 and L19 deduplicate?

File: parlai/core/torch_classifier_agent.py [hunk 11]
```diff
@@ -168,6 +171,158 @@ def value(self) -> float:
...
+    def __init__(
+        self,
+        class_name: Union[int, str],
+        max_bucket_dec_places: int = 3,
+        pos_dict: Optional[Count[float]] = None,
+        neg_dict: Optional[Count[float]] = None,
+    ):
+        # `_pos_dict` keeps track of the probabilities of the positive class
+        self._pos_dict = pos_dict if pos_dict else Counter()
+        # `_neg_dict` keeps track of the probabilities of the positive class
```
Comment (jxmsML) [inline parlai/core/torch_classifier_agent.py:218]: nit typo: the negative class?

**Review by jxmsML — APPROVED** [review]
LGTM!! some nit

File: parlai/core/torch_classifier_agent.py (hunk 11 above)
Comment (liliarose) [reply parlai/core/torch_classifier_agent.py:218]: ooops

File: parlai/scripts/eval_model.py [hunk 12]
```diff
@@ -69,6 +72,21 @@ def setup_args(parser=None):
         default='conversations',
         choices=['conversations', 'parlai'],
     )
+    parser.add_argument(
+        '--area-under-curve-digits',
+        '-auc',
+        type=int,
+        default=-1,
+        help='a positive number indicates to calculate the area under the roc curve and it also determines how many decimal digits of the predictions to keep (higher numbers->more accurate); also used to determine whether or not to calculate the AUC metric',
```
Comment (meganung) [inline parlai/scripts/eval_model.py:80]: nit - more decimal places means more precise instead of accurate
